usb: device_next: make buffers data cache and DMA friendly

jfischer-no opened the pull request:
Use our own version of alloc and unref callbacks to get buffers with specific alignment and granularity. Do not use ref callbac because it breaks alignment.

jfischer-no committed:
usb: device_next: introduce UDC_BUF_POOL_*_DEFINE macros

Introduce UDC_BUF_POOL_*_DEFINE macros based on NET_BUF_POOL_*_DEFINE
macros but use our own version of alloc and unref callbacks to get
buffers with specific alignment and granularity. Also, do not use ref
callback because it breaks alignment.
Also introduces helper macros for defining and checking UDC
driver-compliant static buffers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/udc_common.c:
```diff
@@ -22,7 +22,70 @@ LOG_MODULE_REGISTER(udc, CONFIG_UDC_DRIVER_LOG_LEVEL);
+ * These are appropriate values for nRF54H20, but may be different on other
+ * platforms.
+ */
+#define BUF_ALIGN		32U
```
Why not use `CONFIG_DCACHE_LINE_SIZE` instead? The DWC2 only requires alignment and granularity of 4 in Buffer DMA. The much larger alignment and granularity comes from the D-Cache.

tmon-nordic replied:
Also, should the align and granularity be available in some common header? It would make things simpler for UAC2 API users that do have to provide appropriate audio buffers.

jfischer-no replied:
> Why not use `CONFIG_DCACHE_LINE_SIZE` instead? The DWC2 only requires alignment and granularity of 4 in Buffer DMA. The much larger alignment and granularity comes from the D-Cache.

That is for the the default case that DCACHE is enabled.

> Also, should the align and granularity be available in some common header? It would make things simpler for UAC2 API users that do have to provide appropriate audio buffers.

But it is not up to the UDC or USB subsystem to provide this information, we can have an interface to get it, but the source of this information should be part of the OS. This is the biggest problem, that we lack a consistent source of information about data cache properties and DMA'able areas at the OS level. And CONFIG_DCACHE_LINE_SIZE cannot be trusted because there is also d-cache-line-size' DT property, also there is Kconfig option DCACHE_LINE_SIZE_DETECT.

jfischer-no replied:
I will try to come up with something more usable next week.

jfischer-no committed:
drivers: udc_mcux_ehci: remove bounce buffer

UDC_BUF_POOL_*_DEFINE macro will place buffer pool in __nocache
section if NOCACHE_MEMORY is enabled.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -20,9 +20,7 @@
 #include "usb_device_config.h"
 #include "usb_device_mcux_drv_port.h"
 #include "usb_device_ehci.h"
-#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
```
Is this just some rebase artifact? This commit doesn't seem to belong to this PR.

jfischer-no replied:
I needed it for testing but it was already merged when I pushed.

tmon-nordic commented on include/zephyr/drivers/usb/udc_buf.h:
```diff
@@ -0,0 +1,115 @@
+#define UDC_BUF_GRANULARITY	sizeof(void *)
+#endif
+
+#define UDC_K_HEAP_DEFINE_NO_CACHE(name, size)						\
+	K_HEAP_DEFINE_NOCACHE(name, size)
+
+#define UDC_K_HEAP_DEFINE_DMAABLE(name, size)						\
+	K_HEAP_DEFINE(name, size)
+
+#define UDC_K_HEAP_DEFINE(name, size)							\
+	COND_CODE_1(CONFIG_NOCACHE_MEMORY,						\
+		    (UDC_K_HEAP_DEFINE_NO_CACHE(name, size)),				\
+		    (UDC_K_HEAP_DEFINE_DMAABLE(name, size)))
```
The names are a bit confusing. Both NO_CACHE and DMAABLE has to be in memory region accessible by DMA. The difference is whether the memory has to be invalidated/flushed by D-Cache or not.

For the time being I would go without the intermediate `UDC_K_HEAP_DEFINE_NO_CACHE` and `UDC_K_HEAP_DEFINE_DMAABLE` and just use the `K_HEAP_DEFINE_NOCACHE` and `K_HEAP_DEFINE` directly in the `COND_CODE_1` here.

tmon-nordic replied:
NOCACHE on its own is clear, but because the other one is DMAABLE - it infers that the NOCACHE is not necessarily DMA-able.

tmon-nordic commented on include/zephyr/drivers/usb/udc_buf.h:
```diff
@@ -0,0 +1,115 @@
+#define UDC_BUF_ALIGN		CONFIG_DCACHE_LINE_SIZE
+#define UDC_BUF_GRANULARITY	CONFIG_DCACHE_LINE_SIZE
+#else
+/* Defaults alignment and granularity if there is not data cache */
```
```suggestion
/* Default alignment and granularity to pointer size if there is no data cache */
```

tmon-nordic commented on include/zephyr/drivers/usb/udc_buf.h:
```diff
@@ -0,0 +1,115 @@
+ * @brief Define a new pool for UDC buffers based on fixed-size data
+ *
+ * This macro is similar to `NET_BUF_POOL_VAR_DEFINE`, but provides buffers
+ * with proper alignment and granularity.
```
```suggestion
 * with alignment and granularity suitable for use by UDC driver.
```

tmon-nordic commented on include/zephyr/drivers/usb/udc_buf.h:
```diff
@@ -0,0 +1,115 @@
+ * @param ud_size    User data space to reserve per buffer.
+ * @param fdestroy   Optional destroy callback when buffer is freed.
+ */
+#define UDC_BUF_POOL_VAR_DEFINE(pname, count, size, ud_size, fdestroy)			\
```
Is this replacement for `NET_BUF_POOL_DEFINE`? If yes, then it needs to be used in `usbd_uac2.c` and `usbd_hid_macros.h`

It would be good to have alignment assert in `uac2_buf_alloc()` and `hid_buf_alloc_ext()`, e.g. `__ASSERT(IS_ALIGNED(data, UDC_BUF_ALIGN), "Application provided unaligned buffer")`. The granurality is unfortunately not possible to be checked and has to rely on application developer providing appropriately padded buffers.

tmon-nordic reviewed (changes requested):
I think you can also remove the D-Cache disabled when DMA enabled check in DWC2 driver.

tmon-nordic commented on include/zephyr/drivers/usb/udc_buf.h:
```diff
@@ -0,0 +1,156 @@
+ * Example use to make buffer alignment UDC compliant:
+ *
+ * @code{.c}
+ *     static uint8_t udc_align_attr report[KB_REPORT_COUNT];
```
This example would need the `KB_REPORT_COUNT` to be increased to be multiple of `UDC_BUF_GRANULARITY`, e.g. `ROUND_UP(KB_REPORT_COUNT, UDC_BUF_GRANULARITY)`.

jfischer-no replied:
Yeah, I have played with doxygen, still trying to figure out how to make it better. Maybe some macro to define static and compliant buffer.

dleach02 commented:
I've often thought that it would be useful to have a generic Zephyr API for DMA/cache friend buffer allocation/management. Pretty much every IP block that can support DMA would benefit from this in the presence of a system that also has a cache.

jfischer-no commented:
> I've often thought that it would be useful to have a generic Zephyr API for DMA/cache friend buffer allocation/management. Pretty much every IP block that can support DMA would benefit from this in the presence of a system that also has a cache.
> 
> Would it be useful to use this PR to define the generic API and have the usb use it?

Hopefully not :smiley:, I think it is easy to screw up. This only scratches the surface and is limited to the USB subsystem. What is missing is a general description of data cache and memory, such as which region is DMA'able or which is nocache, on the platform level. Then I do not think we need a new API, we already have too many MM APIs. So e.g. [Shared Multi Heap](https://docs.zephyrproject.org/latest/kernel/memory_management/shared_multi_heap.html) seems to be designed with thought to get appropriate heap based on attributes, but there are missing attributes like to get heap from DMA'able region. I think slabs should also have possibility to select appropriate region, or we go to use only heap, although I do not know performance differences.

dleach02 commented:
I'm using the term "API" a bit loosely. Defining memory regions that are safe for this via DTS would be a way but you would still want some sort of indicator at the application level or device level that can request the memory be located in this area. We have run into this with other drivers and where the buffer being supplied is coming from the application level the application would need some way of allocating memory from a safe region... but in the cases where this has occurred, our observation is that the "contract" in the API doesn't specify if the memory supplied should be safe for DMA/CACHE challenges.

tmon-nordic commented on samples/subsys/usb/uac2_explicit_feedback/src/main.c:
```diff
@@ -36,7 +36,7 @@ LOG_MODULE_REGISTER(uac2_sample, LOG_LEVEL_INF);
  */
 #define I2S_BUFFERS_COUNT   7
-K_MEM_SLAB_DEFINE_STATIC(i2s_tx_slab, MAX_BLOCK_SIZE, I2S_BUFFERS_COUNT, 4);
+K_MEM_SLAB_DEFINE_STATIC(i2s_tx_slab, MAX_BLOCK_SIZE, I2S_BUFFERS_COUNT, UDC_BUF_ALIGN);
```
`MAX_BLOCK_SIZE` here has to be replaced with `ROUND_UP(MAX_BLOCK_SIZE, UDC_BUF_GRANULARITY)`.

jfischer-no replied:
Yes, thank you. Maybe we should look closer at [Shared Multi Heap](https://docs.zephyrproject.org/latest/kernel/memory_management/shared_multi_heap.html) and replace slabs and static buffers with shared_multi_heap_alloc().

tmon-nordic replied:
Maybe, but here the memory slab is imposed by I2S API so the change cannot happen as freely as in the experimental USB stack.

tmon-nordic commented:
> I'm using the term "API" a bit loosely. Defining memory regions that are safe for this via DTS would be a way but you would still want some sort of indicator at the application level or device level that can request the memory be located in this area. We have run into this with other drivers and where the buffer being supplied is coming from the application level the application would need some way of allocating memory from a safe region... but in the cases where this has occurred, our observation is that the "contract" in the API doesn't specify if the memory supplied should be safe for DMA/CACHE challenges.

I think this is best to be handled outside of this PR. I think it would be easiest to have some magic attribute indicator for the function parameter that would indicate that buffer passed in has to be aligned to X and in multiples of Y, but I have absolutely no idea how such thing could be implemented.

dleach02 commented:
I don't disagree. I'm highlighting this because it is a common problem we have seen before.

jfischer-no committed:
drivers: udc_dwc2: enable Internal DMA support by default

Remove BUILD_ASSERT for Internal DMA and DCACHE if both are enabled, and
enable Internal DMA support by default.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

henrikbrixandersen commented:
Is this still supposed to be labelled "DNM"?

hubertmis commented on include/zephyr/drivers/usb/udc_buf.h:
```diff
@@ -0,0 +1,154 @@
+#define Z_UDC_BUF_GRANULARITY	CONFIG_DCACHE_LINE_SIZE
+#else
+/*
+ * Default alignment and granularity to pointer size if the platform does not
```
Why the default is a pointer size, not simply 1 byte (any) alignment?

tmon-nordic replied:
DWC2 Buffer DMA needs 4 bytes, so using 1 byte here would break nRF54H20 USB with D-Cache disabled. Not sure how to properly describe the potential driver DMA alignment requirements.

I believe the buffers given out by default net buf allocator (this PR replaces them) are pointer size aligned.

jfischer-no replied:
This is non-debatable default alignment for all USB controllers and has nothing to do with this PR.

hubertmis replied:
I understand some controllers have alignment requirements (like @tmon-nordic described for DWC2). However, this API seems to be controller-agnostic. That's why I was curious why controller-specific requirements are enforced here. That's OK if that's a common requirement for all supported controllers.

Anyway, it would be good to capture this decision and rationale in a comment to avoid misunderstandings in the future.

jfischer-no replied:
Actually, this is good behavior to deliver the buffer with minimal alignment, this is what k_heap_alloc does. This prevents a lot of problems, not only with USB controllers. I will add a note about alignment requirements when I write documentation about buffer handling in USB.

hubertmis commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -52,37 +52,6 @@ struct udc_mcux_data {
 };
 
-/* TODO: implement the cache maintenance
- * solution1: Use the non-cached buf to do memcpy before/after giving buffer to usb controller.
- * solution2: Use cache API to flush/invalid cache. but it needs the given buffer is
```
What was the reason for selecting "solution1"? Is there some profiling data available comparing the properties of these solutions?

tmon-nordic replied:
Related, but not quite the same: solution2 is implemented in dwc2 otg driver.

jfischer-no replied:
None of these are my solutions, nor was 1 selected. With these changes, the MCUX platform uses nocached region for USB buffers, but the driver can be adapted to use dcache_invalidate/dcache_flush and cached memory region. Sometimes I like to work with former Freescale products, but I do not have time to try it out.¯|_(ツ)_/¯

hubertmis replied:
I see. I reviewed this commit too fast yesterday.

tmon-nordic commented:
Needs rebase due to conflict in `samples/subsys/usb/hid-mouse/src/main.c` (67a31ef2d781231aacb44b6a99681750a01123d5)
